asreview/webapp/api/auth.py

```diff
@@ -0,0 +1,512 @@
+    mailer = Mail(cur_app)
+    # open templates as string and render
+    root_path = Path(cur_app.root_path)
+    with open(root_path / "templates/emails/confirm_account.html", "r") as f:
+        html_text = render_template_string(f.read(), name=name, url=url)
+    with open(root_path / "templates/emails/confirm_account.txt", "r") as f:
+        txt_text = render_template_string(f.read(), name=name, url=url)
+    # create message
+    msg = Message(
+        "ASReview: please confirm your account",
+        recipients=[user.email],
+        sender=config.get("REPLY_ADDRESS"),
+    )
+    msg.body = txt_text
+    msg.html = html_text
+    return mailer.send(msg)
+
+
+# ------------------
+#      ROUTES
+# ------------------
+
+
+@bp.route("/signin", methods=["POST"])
+def signin():
+    email = request.form.get("email").strip()
+    password = request.form.get("password", "")
+
+    # get the user
+    user = User.query.filter(
+        or_(User.identifier == email, User.email == email)
+    ).one_or_none()
+
+    if not user:
+        # user does not exsist
+        result = (404, {"message": f"User account {email} does not exist."})
+    elif not user.confirmed:
+        # account is not confirmed
+        result = (404, {"message": f"User account {email} is not confirmed."})
+    else:
+        # user exists and is confirmed: verify password
+        if user.verify_password(password):
+            logged_in = perform_login_user(user)
+            result = (
+                200,
+                {"logged_in": logged_in, "name": user.get_name(), "id": user.id},
+            )
+        else:
+            # password is wrong
+            if user.origin == "asreview":
+                # if this is an asreview user
+                result = (404, {"message": f"Incorrect password for user {email}"})
+            else:
+                # this must be an OAuth user trying to get in with
+                # a password
+                service = user.origin.capitalize()
+                result = (404, {"message": f"Please login with the {service} service"})
+
+    status, message = result
+    response = jsonify(message)
+    return response, status
+
+
+@bp.route("/signup", methods=["POST"])
+def signup():
+    # this is for the response
+    user_id = False
+
+    # Can we create accounts?
+    if current_app.config.get("ALLOW_ACCOUNT_CREATION", False):
+        email = request.form.get("email", "").strip()
+        name = request.form.get("name", "").strip()
+        affiliation = request.form.get("affiliation", "").strip()
+        password = request.form.get("password")
+        public = bool(int(request.form.get("public", "1")))
+
+        # check if email already exists
+        user = User.query.filter(
+            or_(User.identifier == email, User.email == email)
+        ).one_or_none()
+        # return error if user doesn't exist
+        if isinstance(user, User):
+            result = (403, f'User with email "{email}" already exists.')
```

Comment:
Do we want to give information on whether a user exists or not in these response messages? It will make for an easier UI experience, but maybe less secure. One [source](information: https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#account-creationl) says the message and status_cose should not reveal any information. Do you have an opinion @J535D165 ?